smartsim/_core/utils/timings.py

```diff
@@ -0,0 +1,134 @@
+
+import time
+import typing as t
+from collections import OrderedDict
+
+import numpy as np
+
+from ...log import get_logger
+
+logger = get_logger("PerfTimer")
+
+
+class PerfTimer:
+    def __init__(
+        self,
+        filename: str = "timings",
+        prefix: str = "",
+        timing_on: bool = True,
+        debug: bool = False,
+    ):
+        self._start: t.Optional[float] = None
+        self._interm: t.Optional[float] = None
+        self._timings: OrderedDict[str, list[t.Union[float, int, str]]] = OrderedDict()
+        self._timing_on = timing_on
+        self._filename = filename
+        self._prefix = prefix
+        self._debug = debug
+
+    def _add_label_to_timings(self, label: str) -> None:
+        if label not in self._timings:
+            self._timings[label] = []
+
+    @staticmethod
+    def _format_number(number: float | int) -> str:
+        return f"{number:0.4e}"
+
+    def start_timings(
+        self,
+        first_label: t.Optional[str] = None,
+        first_value: t.Optional[float | int] = None,
+    ) -> None:
+        if self._timing_on:
+            if first_label is not None and first_value is not None:
+                mod_label = self._make_label(first_label)
+                value = self._format_number(first_value)
+                self._log(f"Started timing: {first_label}: {value}")
+                self._add_label_to_timings(mod_label)
+                self._timings[mod_label].append(value)
+            self._start = time.perf_counter()
+            self._interm = time.perf_counter()
+
+    def end_timings(self) -> None:
+        if self._timing_on and self._start is not None:
+            mod_label = self._make_label("total_time")
+            self._add_label_to_timings(mod_label)
+            delta = self._format_number(time.perf_counter() - self._start)
+            self._timings[self._make_label("total_time")].append(delta)
+            self._log(f"Finished timing: {mod_label}: {delta}")
+            self._interm = None
+
+    def _make_label(self, label: str) -> str:
+        return self._prefix + label
+
+    def _get_delta(self) -> float | int:
+        if self._interm is None:
+            return 0
+        return time.perf_counter() - self._interm
+
+    def get_last(self, label: str) -> str:
+        mod_label = self._make_label(label)
+        if mod_label in self._timings:
+            value = self._timings[mod_label][-1]
+            if value:
+                return f"{label}: {value}"
+
+        return "Not measured yet"
+
+    def measure_time(self, label: str) -> None:
+        if self._timing_on and self._interm is not None:
+            mod_label = self._make_label(label)
+            self._add_label_to_timings(mod_label)
+            delta = self._format_number(self._get_delta())
+            self._timings[mod_label].append(delta)
+            self._log(f"{mod_label}: {delta}")
+            self._interm = time.perf_counter()
+
+    def _log(self, msg: str) -> None:
```

Comment:
Consider configuring a specifically-named logger, instead. You can configure externally and avoid conditionals like `if self._debug:` by making use of the log level. this will also reduce the need to do formatting in code (by using log format) and passing in file names, `timing_on`, etc.

Reply:
Again, I'll post a `PerfTimer` overhaul ticket.